### ISSUE TITLE
Parity between the different JSS models

### ISSUE THREAD

**Issue by MysterAitch** [javaparser/javaparser#3192]
At the moment we have multiple models within the symbol solver (iirc the reflection model, the jar one, and the javaparser (source code) one).

Not all areas are consistent with each other -- often rightly so -- but a quick audit doing a side by side comparison between the three would be helpful to identify any inconsistencies.

Documentation that cross references between them might be good too, but additional thought is needed to have a strong view on this.

**Comment by deadlocklogic** [comment]
Indeed, there are also many missing helpful methods which can be added to the API.
Btw I was studying `javaparser-solver` module a bit because I was dealing with many issues which I was trying to fix. There are plenty of weaknesses with the generic and anonymous classes resolution parts which need drastic changes. I made tons of various tests which fail. I believe the first step to write more rigorous tests/fixes starts with few key points.
1. Naming local and anonymous class with its name generated by the JRE. Because currently an anonymous class name after getting resolved returns `"Anonymous-" + UUID.randomUUID()` which is kind of confusing/not reproducible, unless involving more digging while using the `toAst()` method again which introduces more tests/breaks the parity between the JSS models.
2. Sticking with Java/Javaassist conventions: using getName()/getSimpleName() instead of getQualifiedName()/getName() (more intuitive).
3. Unified resolution models with a tight API.
4. Many navigating methods in the `JavaParserFacade` can be converted automatically to static and can be safely moved into the `Navigator` class.
And many more room for improvement exist.

**Comment by deadlocklogic** [comment]
Actually, I am including many of my own snippets of my side projects in my fixes (as I mentioned before, I am building a java transpiler and a refactoring tool). So I already got an almost drop-in-place PR for the first issue.
For instance imagine this situation:
```java
public class Program {
    class InnerClass1 {
    }

    class InnerClass2 {
    }

    {
        new InnerClass1() {
            class InnerClass2 {
            }
        };
    }
}
```
The `InnerClass2` class inside the `ObjectCreationExpr` body `getFullyQualifiedName()` is `Program.InnerClass2` and insanely wrong!!! Which invites all hell of bugs by having the same qualified name as the upper class `InnerClass2` in the hierarchy.

**Comment by jlerbsc** [comment]
I think this example is an illustration of a bug in JP but if you put this description in this issue it will be difficult to manage this issue.
The actual problem in this case is that the TypeDeclaration.getFullyQualifiedName() try to compose the fqn from TypeDeclaration ancestor. The comment is explicit "Returns the fully qualified name of this type, derived only from information available in this compilation unit. (**So no symbol solving happens**.)"

**Comment by deadlocklogic** [comment]
Yes, I know that it is a bug, in fact there are plenty of bugs of this kind that I am facing mainly related to generic/anonymous and `ObjectCreationExpr` parts where the solver is relatively weak.
Objectively speaking the `new InnerClass1()` type in this case should be named `Program$1` and the inner class inside it should be named `Program$1$InnerClass2` not `"Anonymous-" + UUID.randomUUID()` and `Program.InnerClass2` respectively, this is actually what the java runtime generates as synthetic classes.
Actually, replying to **"no symbol solving happens"**, when trying to get the `fullQualifiedName` of any class in the `CompilationUnit`, we traverse all the hierarchy from bottom up to find it. Same thing here, we do not need to solve anything to generate a name for anonymous/local classes, it just a matter of traversing the whole hierarchy  and calculate the names accordingly, a simple visitor can do it, and I already did this in my other projects.
And just a question, why not sticking with **java/javassist** conventions and use `getSimpleName()/getName()/getCanonicaleName()` instead of `getName()/getFullyQualifiedName()`? What is the point of replacing '$' with '.' the whole time across the API?

**Comment by MysterAitch** [comment]
I cannot comment on it all, but I suspect that some of it will likely the result of "history" -- with different functionality being implemented at different times e.g. adding javassist later, once everything else was already implemented.

**Comment by deadlocklogic** [comment]
Yes most likely.
Still there is a major part which I missed talking about and which is immensely needed: improved documentations for symbolsolver, especially private methods because fixing a single bug might take days with the current situation. You need to dig every other methods (which most of them have similar naming and sometimes different functionality) and try to figure out what is the difference between each other every time.
